sql: panic due to inconsistent descriptor state

https://teamcity.cockroachdb.com/viewLog.html?buildId=38866&tab=buildResultsDiv&buildTypeId=Cockroach_Nightlies_RandomSyntaxTests is a very interesting test failure. Stack trace copied below.

The error appears to be a failed call to `desc.GetTable()`, but the result isn't checked since it is assumed (due to the case switch above it) that everything already has the correct type. I looked into it but couldn't figure out what was going on. I also attempted to reproduce locally but was unable after running RSG for an hour or so.

My current best explanations are:

1. Some transient bit flip error or some other hardware-related thing.
1. A race condition where a descriptor was being renamed or a table dropped and a database added or something in another goroutine.

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: ALTER INDEX IF EXISTS BIGINT RENAME TO TREAT: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: ALTER INDEX IF EXISTS BIGINT RENAME TO TREAT: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1b4 pc=0x98be42]

goroutine 22024 [running]:
panic(0x16b8660, 0xc420f2ad00)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4201c2000)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x16b8660, 0xc420f2ad00)
    /usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc424f5d7d5, 0x2c)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:414 +0x149
panic(0x16db680, 0xc420012130)
    /usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*TableDescriptor).MaybeUpgradeFormatVersion(0x0, 0x1784480)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/structured.go:345 +0x22
github.com/cockroachdb/cockroach/pkg/sql.(*planner).getDescriptor(0xc42685b2f8, 0x2325920, 0xc42159c280, 0x2337780, 0xc424a87440, 0x0, 0x6, 0x74)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/descriptor.go:211 +0x2b3
github.com/cockroachdb/cockroach/pkg/sql.(*planner).getTableOrViewDesc(0xc42685b2f8, 0xc423399180, 0x8b1845, 0xc4216bb6a0, 0x6)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/table.go:146 +0x1e0
github.com/cockroachdb/cockroach/pkg/sql.(*planner).getTableDesc(0xc42685b2f8, 0xc423399180, 0x942260, 0x2319b60, 0x2)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/table.go:158 +0x39
github.com/cockroachdb/cockroach/pkg/sql.(*planner).mustGetTableDesc(0xc42685b2f8, 0xc423399180, 0xc4216bb6a0, 0x6, 0x2)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/table.go:197 +0x39
github.com/cockroachdb/cockroach/pkg/sql.(*planner).findTableContainingIndex(0xc42685b2f8, 0xc425bcb045, 0x5, 0xc424f5d7eb, 0x6, 0x0, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/table.go:556 +0x13d
github.com/cockroachdb/cockroach/pkg/sql.(*planner).expandIndexName(0xc42685b2f8, 0xc422bf3860, 0x5d1010, 0x27, 0x22a3ae16)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/table.go:585 +0xaa
github.com/cockroachdb/cockroach/pkg/sql.(*planner).RenameIndex(0xc42685b2f8, 0xc42208fa60, 0xc4255a0f00, 0xc423e4c258, 0xf07282, 0xc4208ec100)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/rename.go:254 +0x53
github.com/cockroachdb/cockroach/pkg/sql.(*planner).newPlan(0xc42685b2f8, 0x7f7d1fd13bf8, 0xc42208fa60, 0x0, 0x0, 0x0, 0xc423957301, 0x48, 0x0, 0xecfa8b944, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:287 +0x1f3
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc42685b2f8, 0x7f7d1fd13bf8, 0xc42208fa60, 0x845401, 0xc42667eaa0, 0xc4ffffffff, 0x0, 0x2)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:219 +0x70
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4202b7a40, 0x7f7d1fd13bf8, 0xc42208fa60, 0xc42685b2f8, 0xc423e4c501, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1151 +0xc2
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4202b7a40, 0x7f7d1fd13bf8, 0xc42208fa60, 0xc42685b2f8, 0x101, 0xc42685b268, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1042 +0x2c0
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc4202b7a40, 0xc421925a80, 0x1, 0x1, 0xc42685b2f8, 0xc42685b268, 0x101, 0x6, 0xc423e4cf58, 0x60ca4a, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:789 +0x34f
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4202b7a40, 0xc42685b2f8, 0x1, 0xc42685b268, 0xc421925b10, 0xc421925a80, 0x1, 0x1, 0x0, 0x0, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:714 +0xeb
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func2(0xc424cf34a0, 0xc421925b10, 0x4000000000000000, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:576 +0x22f
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc424cf34a0, 0xc423b40101, 0xc4200191c0, 0xc423b4f0e0, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:518 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4202b7a40, 0xc42685b200, 0xc424f5d7d5, 0x2c, 0x0, 0x0, 0x0, 0x0, 0xc423e4d700)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:590 +0x572
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4202b7a40, 0xc42685b200, 0xc424f5d7d5, 0x2c, 0x0, 0x0, 0x0, 0x0, 0xc424f5d700)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:420 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc420a88780, 0x7f7d1fde92c0, 0xc4256b7050, 0xc424f5d7d5, 0x2c, 0x0, 0x0, 0x0, 0x0, 0xaca801, ...)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:654 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc420a88780, 0x7f7d1fde92c0, 0xc4256b7050, 0xc420a887a8, 0x31, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:328 +0xb3
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc420a88780, 0x7f7d1fde92c0, 0xc4256b7050, 0xc42667eac0, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:280 +0x820
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc42019bb00, 0x7f7d1fde92c0, 0xc4256b7050, 0x2336620, 0xc4241ad880, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:245 +0x63e
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x2336680, 0xc425a44790)
    /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:488 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4201c2000, 0xc4201701b8, 0x2336680, 0xc425a44790, 0xc420554020)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
    /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x277
```

--- a/pkg/sql/descriptor.go
+++ b/pkg/sql/descriptor.go
@@ -208,15 +208,15 @@ func (p *planner) getDescriptor(
 	switch t := descriptor.(type) {
 	case *sqlbase.TableDescriptor:
 		table := desc.GetTable()
+		if table == nil {
+			return false, errors.Errorf("%q is not a table", plainKey.Name())
+		}
 		table.MaybeUpgradeFormatVersion()
 		// TODO(dan): Write the upgraded TableDescriptor back to kv. This will break
 		// the ability to use a previous version of cockroach with the on-disk data,
 		// but it's worth it to avoid having to do the upgrade every time the
 		// descriptor is fetched. Our current test for this enforces compatibility
 		// backward and forward, so that'll have to be extended before this is done.
-		if table == nil {
-			return false, errors.Errorf("%q is not a table", plainKey.Name())
-		}
 		if err := table.Validate(p.txn); err != nil {
 			return false, err
 		}

